### DISCUSSION TITLE
Generating training data from trajectory

### DISCUSSION THREAD

**Discussion by jonathan-booth** [mir-group/allegro#74] [Q&A]
I have a few trajectories where, at each frame, I have a text file where each line is an atom and we have values for the atom type, the x,y,z positions, and the x,y,z components of the force vector. Each text file also has the total potential energy. I have trajectories for different sized systems. How do I convert these data into something that Nequip (Allegro really) can work with? I've read the docs but not found an answer sorry if I missed something obvious.

Thanks in advance

**Comment by DavidW99** [comment]
Dear Jonathan,

You can try to convert your file to `extxyz` format using `ase` package. Reference: https://wiki.fysik.dtu.dk/ase/ase/io/io.html. Systems with different cell sizes are OK to be put together as your data set. But, please check if the boundary condition (PBC or not PBC) of your output xyz is correctly set because it can affect the result. 

Then in your yaml file, you can use the following to read your data file.
```
dataset: ase                                                                   
dataset_file_name:  your_data.xyz
ase_args:
  format: extxyz
```

**Reply by jonathan-booth** [reply]
Thanks. I didn't realise ase could do this with data from some other simulation I thought it was just for the output of simulations that were run with ase itself. It makes sense now thanks,

**Reply by DavidW99** [reply]
No problem. ASE can read its supported file format. But, I realized that ASE probably cannot read txt directly. If that is the case, you can read line-by-line or `numpy.loadtxt()` to extract relevant info and construct the `xyz` file using `ase`. Then, you can streamline the process for all your trajectories.

**Reply by jonathan-booth** [reply]
Thanks, that makes sense. Do you have an example of a training dataset in extxyz format just so I can check that mine is done correctly? Let's say I have 2 atoms and 3 frames in the trajectory, and I have the position, force and potential energy at each trajectory frame, then if my understanding is correct my data.xyz file should look like this:

2
Lattice="A 0.0 0.0 0.0 B 0.0 0.0 0.0 C" Properties=species:S:1:pos:R:3:F:R:3 Time=0.0 Energy=E0
C       x     y    z      Fx    Fy     Fz
C       x     y    z      Fx    Fy     Fz

Lattice="A 0.0 0.0 0.0 B 0.0 0.0 0.0 C" Properties=species:S:1:pos:R:3:F:R:3 Time=1.0 Energy=E1
C       x     y    z      Fx    Fy     Fz
C       x     y    z      Fx    Fy     Fz

Lattice="A 0.0 0.0 0.0 B 0.0 0.0 0.0 C" Properties=species:S:1:pos:R:3:F:R:3 Time=2.0 Energy=E2
C       x     y    z      Fx    Fy     Fz
C       x     y    z      Fx    Fy     Fz


Where E0, E1 and E2 are the values of the potential energy at each timestep. And x,y,z,Fx,Fy,Fz are the position and force vectors for each atom at each timestep.

Here I have all the trajectory frames in the same .xyz file, is this correct? And if so is it correct to repeat the Lattice and Properties line at every timestep? 

Thanks

**Reply by DavidW99** [reply]
No problem, the example is given in this `ase` page https://wiki.fysik.dtu.dk/ase/ase/io/formatoptions.html#extxyz.

This is an example format with forces, which is a per-atom value listed in each row:

```
Lattice="1.0 0.0 0.0 0.0 2.0 0.0 0.0 0.0 3.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-100 stress="0.0002 0.003 -0.004 0.003 -0.004 -0.005 -0.004 -0.005 -0.006" free_energy=-100 pbc="T T T"
H      14.80702000       9.47114000      14.83362000       0.05630700       0.18844700       0.12262800
C      14.40303000       9.58896000       4.81541000       0.28543000      -0.56192100       0.59003500
Al       0.15995000       9.61979000       9.87331000      -0.15746000       0.51516600       0.14365000
```

I suggest you to create an instance `atom` from Atom object following https://wiki.fysik.dtu.dk/ase/ase/atoms.html#list-of-all-methods and do `atom.new_array(‘forces’, your_forces_array)` to add forces.

**Reply by jonathan-booth** [reply]
OK thanks. How does the extxyz format work with a trajectory? I have one molecule with different positions and forces at each timestep, and I would like to learn force as a function of atomic positions. So do I have a separate block for each frame in the same extxyz file as in my post above or do I have a separate file for each frame? And sorry about the other question on the Nequip page being a duplicate of this, I posted here and then thought it would be better on the Nequip page and I must have forgotten to delete this one.

**Reply by DavidW99** [reply]
No worries! Indeed, it will be many blocks of frames. For a trajectory, extxyz contains a stack of single frame and the frames are separated by the number of atoms in the previous frame shown below.
```
3
Lattice="1.0 0.0 0.0 0.0 2.0 0.0 0.0 0.0 3.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-100 stress="0.0002 0.003 -0.004 0.003 -0.004 -0.005 -0.004 -0.005 -0.006" free_energy=-100 pbc="T T T"
H      14.80702000       9.47114000      14.83362000       0.05630700       0.18844700       0.12262800
C      14.40303000       9.58896000       4.81541000       0.28543000      -0.56192100       0.59003500
Al       0.15995000       9.61979000       9.87331000      -0.15746000       0.51516600       0.14365000
2
Lattice="1.0 0.0 0.0 0.0 2.0 0.0 0.0 0.0 3.0" Properties=species:S:1:pos:R:3:forces:R:3 energy=-100 stress="0.0002 0.003 -0.004 0.003 -0.004 -0.005 -0.004 -0.005 -0.006" free_energy=-100 pbc="T T T"
H      14.80702000       9.47114000      14.83362000       0.05630700       0.18844700       0.12262800
C      14.40303000       9.58896000       4.81541000       0.28543000      -0.56192100       0.59003500
```

**Reply by KP-303** [reply]
Thanks! One final question, it says in the ase docs that an extxyz file must contain the lattice vectors. What if I don't have periodic boundaries? If it's just some atoms in the gas phase in vacuum for instance. Do I just put very large lattice vectors in the file so there will be no periodic images within the cutoff during training?

**Reply by DavidW99** [reply]
Great question. For gas phase, you can set `pbc` to be False and leave `cell` parameter in the `ase.Atoms` object as default. Then, the feature vector will not take the periodic image of atoms. I think `ase.Atom` object does not require setting the `cell`, i.e. lattice vectors. Please refer to https://wiki.fysik.dtu.dk/ase/ase/atoms.html#list-of-all-methods for more info. 

However, you should be aware that there is no cell stress for the gas phase, and I think this discussion might be helpful https://github.com/mir-group/allegro/discussions/34.
